### PR TITLE
Create `User` and `Breadcrumb` from map

### LINKED ISSUE #2534: Create `User` and `Breadcrumb` data classes from a Map
We have a lot of duplicate code across Hybrid SDKs, as most of them need to create the native data classes from a map.

- https://github.com/getsentry/team-mobile/issues/59

The goal is to add the following:
`User.fromMap(Map<String, Any> user)`
`Breadcrumb.fromMap(Map<String, Any> crumb)`

As all the hybrid SDKs can use this.

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Create `User` and `Breadcrumb` from map.

## :bulb: Motivation and Context

Closes #2534

## :green_heart: How did you test it?

Added unit tests.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry/src/main/java/io/sentry/Breadcrumb.java [hunk 1]
```diff
@@ -59,6 +60,25 @@ public Breadcrumb(final @NotNull Date timestamp) {
...
+   * Creates breadcrumb from a map.
+   *
+   * @param map - The breadcrumb data as map
+   * @param options - the sentry options
+   * @return the breadcrumb
+   */
+  public static @Nullable Breadcrumb fromMap(
+      @NotNull Map<String, Object> map, @NotNull SentryOptions options) {
+    try {
+      String json = options.getSerializer().serialize(map);
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/Breadcrumb.java:73]: tbqh, I'm unsure about this approach with ser/deser roundtrip. Is there a specific reason for that, other than less code? I'd prefer just a simple approach with iterating over key-values and having if-checks for each key to construct a breadcrumb.

With the current approach, we potentially also need to call this from a background thread on the consumer's side

Comment (denrase) [reply sentry/src/main/java/io/sentry/Breadcrumb.java:73]: Yeah, the reason was to leverage existing (tested) code to do the deser. It is runtime overhead though, i will change as suggested.

File: sentry/src/main/java/io/sentry/protocol/User.java [hunk 2]
```diff
@@ -57,6 +59,80 @@ public User(final @NotNull User user) {
...
+          user.ipAddress = (value instanceof String) ? (String) value : null;
+          break;
+        case JsonKeys.DATA:
+          final Map<Object, Object> data =
+              (value instanceof Map) ? (Map<Object, Object>) value : null;
+          if (data != null) {
+            final ConcurrentHashMap<String, String> userData = new ConcurrentHashMap<>();
+            for (Map.Entry<Object, Object> dataEntry : data.entrySet()) {
+              if ((dataEntry.getKey() instanceof String)
+                  && (dataEntry.getValue() instanceof String)) {
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/protocol/User.java:98]: I know user's data is `<String,String>` for some reason, but what if lift up this instanceof check, and instead just call `dataEntry.getValue().toString()` if it's not null? Same for `other`.

I guess this way it's more flexible for the hybrid sdks to set whatever value in the data bag (except objects, but the object we have no idea how to handle anyway). We could also add more javadoc about that behavior.

**Review by romtsn — APPROVED** [review]
lgtm, thanks 👍 just a reminder to also add handling for GEO and NAME after your other PR is merged in:

- [ ] Add handling for `Geo` and `Name` of `User`

File: sentry/src/main/java/io/sentry/protocol/User.java (hunk 2 above)
Comment (denrase) [reply sentry/src/main/java/io/sentry/protocol/User.java]: Updated the handling and added the missing fields. Would you add the doc directly to the new static methods?

Comment (romtsn) [reply sentry/src/main/java/io/sentry/protocol/User.java]: yeah, that should be enough

Comment (denrase) [reply sentry/src/main/java/io/sentry/protocol/User.java]: Updated. If this is alright with you we should be ready to merge.

**Comment by denrase** [comment]
Are we rdy we merge or is anything missing here?

**Comment by romtsn** [comment]
> Are we rdy we merge or is anything missing here?

the CI is red, but good to merge otherwise :)

**Comment by denrase** [comment]
@romtsn There seem to be some timeouts on CI which should not relate to the code. Are those known?

**Comment by romtsn** [comment]
> @romtsn There seem to be some timeouts on CI which should not relate to the code. Are those known?

yeah it's just saucelabs being flaky.. I'm gonna merge this

**Pull request merged**